File: Core/ResultsManager.cs
using PointerFinder2.DataModels;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace PointerFinder2.Core
{
    // Manages the state of the scan results, including data, sorting, and undo/redo history.
    public class ResultsManager
    {
        public IReadOnlyList<PointerPath> CurrentResults => _currentResults;
        public int ResultsCount => _currentResults.Count;
        public bool HasResults => _currentResults.Any();
        public bool CanUndo => _undoStack.Any();
        public bool CanRedo => _redoStack.Any();

        public string SortedColumnName { get; private set; }
        public SortOrder SortOrder { get; private set; } = SortOrder.None;

        private List<PointerPath> _currentResults = new List<PointerPath>();
        private readonly Stack<List<PointerPath>> _undoStack = new Stack<List<PointerPath>>();
        private readonly Stack<List<PointerPath>> _redoStack = new Stack<List<PointerPath>>();

        public event EventHandler<ResultsChangedEventArgs> ResultsChanged;
        public event EventHandler<SortChangedEventArgs> SortChanged;

        public void SetNewResults(List<PointerPath> newResults, bool isNewDataSet)
        {
            if (isNewDataSet)
            {
                PushUndoState();
            }
            _currentResults = newResults ?? new List<PointerPath>();
            if (isNewDataSet)
            {
                ClearSorting();
            }
            ResultsChanged?.Invoke(this, new ResultsChangedEventArgs(isNewDataSet));
        }

        public void ClearResults()
        {
            PushUndoState();
            _currentResults.Clear();
            ClearSorting();
            ResultsChanged?.Invoke(this, new ResultsChangedEventArgs(true));
        }

        public void Sort(string columnName, SortOrder currentSortOrder)
        {
            if (!HasResults) return;

            SortOrder newOrder;
            if (SortedColumnName == columnName)
            {
                newOrder = (currentSortOrder == SortOrder.Ascending) ? SortOrder.Descending : SortOrder.Ascending;
            }
            else
            {
                newOrder = SortOrder.Ascending;
            }

            _currentResults.Sort((p1, p2) =>
            {
                int compareResult = 0;
                if (columnName == "colBase") compareResult = p1.BaseAddress.CompareTo(p2.BaseAddress);
                else if (columnName == "colFinal") compareResult = p1.FinalAddress.CompareTo(p2.FinalAddress);
                else if (columnName.StartsWith("colOffset"))
                {
                    int offsetIndex = int.Parse(columnName.Replace("colOffset", "")) - 1;
                    int offset1 = (offsetIndex < p1.Offsets.Count) ? p1.Offsets[offsetIndex] : int.MinValue;
                    int offset2 = (offsetIndex < p2.Offsets.Count) ? p2.Offsets[offsetIndex] : int.MinValue;
                    compareResult = offset1.CompareTo(offset2);
                }
                return (newOrder == SortOrder.Descending) ? -compareResult : compareResult;
            });

            SortedColumnName = columnName;
            SortOrder = newOrder;

            SortChanged?.Invoke(this, new SortChangedEventArgs(columnName, newOrder));
            ResultsChanged?.Invoke(this, new ResultsChangedEventArgs(false));
        }

        public async Task SortByLowestOffsetsAsync()
        {
            if (!HasResults) return;

            PushUndoState();
            ClearSorting();

            await Task.Run(() =>
            {
                var offsetCache = _currentResults.ToDictionary(
                    p => p,
                    p => p.Offsets.Select(o => Math.Abs(o)).OrderBy(o => o).ToList()
                );

                _currentResults.Sort((p1, p2) =>
                {
                    if (GlobalSettings.SortByLevelFirst)
                    {
                        int levelComparison = p1.Offsets.Count.CompareTo(p2.Offsets.Count);
                        if (levelComparison != 0) return levelComparison;
                    }

                    var p1Offsets = offsetCache[p1];
                    var p2Offsets = offsetCache[p2];

                    int minCount = Math.Min(p1Offsets.Count, p2Offsets.Count);
                    for (int i = 0; i < minCount; i++)
                    {
                        int comparison = p1Offsets[i].CompareTo(p2Offsets[i]);
                        if (comparison != 0) return comparison;
                    }

                    return p1.Offsets.Count.CompareTo(p2.Offsets.Count);
                });
            });

            ResultsChanged?.Invoke(this, new ResultsChangedEventArgs(false));
        }

        // Added a new method to handle the result of a filtering operation, ensuring an undo state is created.
        public void ApplyFilterResults(List<PointerPath> filteredPaths)
        {
            PushUndoState();
            _currentResults = filteredPaths ?? new List<PointerPath>();
            ResultsChanged?.Invoke(this, new ResultsChangedEventArgs(isNewDataSet: false));
        }

        public void DeleteSelected(IEnumerable<int> indicesToRemove)
        {
            PushUndoState();
            var indicesSet = new HashSet<int>(indicesToRemove);
            _currentResults = _currentResults.Where((path, index) => !indicesSet.Contains(index)).ToList();
            ResultsChanged?.Invoke(this, new ResultsChangedEventArgs(false));
        }

        public void Undo()
        {
            if (!CanUndo) return;
            _redoStack.Push(new List<PointerPath>(_currentResults));
            _currentResults = _undoStack.Pop();
            ResultsChanged?.Invoke(this, new ResultsChangedEventArgs(false));
        }

        public void Redo()
        {
            if (!CanRedo) return;
            _undoStack.Push(new List<PointerPath>(_currentResults));
            _currentResults = _redoStack.Pop();
            ResultsChanged?.Invoke(this, new ResultsChangedEventArgs(false));
        }

        public void ClearHistory()
        {
            _undoStack.Clear();
            _redoStack.Clear();
        }

        public void ClearSorting()
        {
            SortedColumnName = null;
            SortOrder = SortOrder.None;
            SortChanged?.Invoke(this, new SortChangedEventArgs(null, SortOrder.None));
        }

        private void PushUndoState()
        {
            _undoStack.Push(new List<PointerPath>(_currentResults));
            _redoStack.Clear();
        }
    }

    public class ResultsChangedEventArgs : EventArgs
    {
        public bool IsNewDataSet { get; }
        public ResultsChangedEventArgs(bool isNewDataSet)
        {
            IsNewDataSet = isNewDataSet;
        }
    }

    public class SortChangedEventArgs : EventArgs
    {
        public string ColumnName { get; }
        public SortOrder SortOrder { get; }
        public SortChangedEventArgs(string columnName, SortOrder sortOrder)
        {
            ColumnName = columnName;
            SortOrder = sortOrder;
        }
    }
}

File: src/Core/ResultsManager.cpp
#include "ResultsManager.h"

#include "../Models/GlobalSettings.h"

#include <algorithm>
#include <QtConcurrent>
#include <unordered_map>

namespace PointerFinder2::Core {

    using namespace PointerFinder2::DataModels;

    ResultsManager::ResultsManager(QObject* parent) : QObject(parent) {}

    ResultsManager::~ResultsManager() {
        m_sortWatcher.cancel();
        m_sortWatcher.waitForFinished();
    }

    const std::vector<PointerPath>& ResultsManager::getCurrentResults() const {
        return m_currentResults;
    }

    int ResultsManager::getResultsCount() const {
        return static_cast<int>(m_currentResults.size());
    }

    bool ResultsManager::hasResults() const {
        return !m_currentResults.empty();
    }

    bool ResultsManager::canUndo() const {
        return !m_undoStack.empty();
    }

    bool ResultsManager::canRedo() const {
        return !m_redoStack.empty();
    }

    QString ResultsManager::getSortedColumnName() const {
        return m_sortedColumnName;
    }

    Qt::SortOrder ResultsManager::getSortOrder() const {
        return m_sortOrder;
    }

    void ResultsManager::setNewResults(std::vector<PointerPath> newResults, bool isNewDataSet) {
        if (isNewDataSet) {
            pushUndoState();
        }
        m_currentResults = std::move(newResults);
        if (isNewDataSet) {
            clearSorting();
        }
        emit resultsChanged(isNewDataSet);
    }

    // Wipes out the active list of found pointer paths, saving the previous state to the undo history.
    void ResultsManager::clearResults() {
        pushUndoState();
        m_currentResults.clear();
        clearSorting();
        emit resultsChanged(true);
    }

    void ResultsManager::sort(const QString& columnName, Qt::SortOrder currentSortOrder) {
        if (!hasResults()) return;

        Qt::SortOrder newOrder = (m_sortedColumnName == columnName)
            ? ((currentSortOrder == Qt::AscendingOrder) ? Qt::DescendingOrder : Qt::AscendingOrder)
            : Qt::AscendingOrder;

        int offsetIndex = -1;
        bool isBase = false;
        bool isFinal = false;

        if (columnName == "colBase") {
            isBase = true;
        }
        else if (columnName == "colFinal") {
            isFinal = true;
        }
        else if (columnName.startsWith("colOffset")) {
            offsetIndex = columnName.mid(9).toInt() - 1;
        }

        std::sort(m_currentResults.begin(), m_currentResults.end(), [isBase, isFinal, offsetIndex, newOrder](const PointerPath& p1, const PointerPath& p2) {
            if (isBase) {
                return (newOrder == Qt::DescendingOrder) ? (p1.baseAddress > p2.baseAddress) : (p1.baseAddress < p2.baseAddress);
            }
            else if (isFinal) {
                return (newOrder == Qt::DescendingOrder) ? (p1.finalAddress > p2.finalAddress) : (p1.finalAddress < p2.finalAddress);
            }
            else if (offsetIndex >= 0) {
                int offset1 = (offsetIndex < static_cast<int>(p1.offsets.size())) ? p1.offsets[offsetIndex] : std::numeric_limits<int>::min();
                int offset2 = (offsetIndex < static_cast<int>(p2.offsets.size())) ? p2.offsets[offsetIndex] : std::numeric_limits<int>::min();
                return (newOrder == Qt::DescendingOrder) ? (offset1 > offset2) : (offset1 < offset2);
            }
            return false;
            });

        m_sortedColumnName = columnName;
        m_sortOrder = newOrder;

        emit sortChanged(columnName, newOrder);
        emit resultsChanged(false);
    }

    void ResultsManager::sortByLowestOffsets() {
        if (!hasResults()) return;

        m_sortWatcher.cancel();
        m_sortWatcher.waitForFinished();

        pushUndoState();
        clearSorting();

        std::vector<PointerPath> resultsCopy = m_currentResults;
        bool sortByLevelFirst = GlobalSettings::sortByLevelFirst;

        QFuture<void> future = QtConcurrent::run([this, resultsCopy = std::move(resultsCopy), sortByLevelFirst]() mutable {
            std::vector<std::vector<int32_t>> offsetCache(resultsCopy.size());
            std::vector<size_t> indices(resultsCopy.size());

            for (size_t i = 0; i < resultsCopy.size(); ++i) {
                indices[i] = i;
                auto& absOffsets = offsetCache[i];
                absOffsets.reserve(resultsCopy[i].offsets.size());
                for (int32_t o : resultsCopy[i].offsets) {
                    absOffsets.push_back(std::abs(o));
                }
                std::sort(absOffsets.begin(), absOffsets.end());
            }

            std::sort(indices.begin(), indices.end(), [&](size_t i1, size_t i2) {
                const auto& p1 = resultsCopy[i1];
                const auto& p2 = resultsCopy[i2];

                if (sortByLevelFirst) {
                    if (p1.offsets.size() != p2.offsets.size()) {
                        return p1.offsets.size() < p2.offsets.size();
                    }
                }

                const auto& p2Offsets = offsetCache[i2];
                const auto& p1Offsets = offsetCache[i1];

                size_t minCount = std::min(p1Offsets.size(), p2Offsets.size());
                for (size_t i = 0; i < minCount; ++i) {
                    if (p1Offsets[i] != p2Offsets[i]) {
                        return p1Offsets[i] < p2Offsets[i];
                    }
                }

                return p1.offsets.size() < p2.offsets.size();
                });

            std::vector<PointerPath> sortedResults;
            sortedResults.reserve(resultsCopy.size());
            for (size_t idx : indices) {
                sortedResults.push_back(std::move(resultsCopy[idx]));
            }

            QMetaObject::invokeMethod(this, [this, sortedResults = std::move(sortedResults)]() mutable {
                m_currentResults = std::move(sortedResults);
                emit resultsChanged(false);
                emit asyncSortCompleted();
                }, Qt::QueuedConnection);
            });

        m_sortWatcher.setFuture(future);
    }

    void ResultsManager::applyFilterResults(const std::vector<PointerPath>& filteredPaths) {
        pushUndoState();
        m_currentResults = filteredPaths;
        emit resultsChanged(false);
    }

    void ResultsManager::deleteSelected(const std::vector<int>& indicesToRemove) {
        if (indicesToRemove.empty()) return;
        pushUndoState();

        std::vector<bool> removeMap(m_currentResults.size(), false);
        for (int idx : indicesToRemove) {
            if (idx >= 0 && idx < static_cast<int>(m_currentResults.size())) {
                removeMap[idx] = true;
            }
        }

        std::vector<PointerPath> filtered;
        filtered.reserve(m_currentResults.size() - indicesToRemove.size());
        for (size_t i = 0; i < m_currentResults.size(); ++i) {
            if (!removeMap[i]) {
                filtered.push_back(m_currentResults[i]);
            }
        }

        m_currentResults = std::move(filtered);
        emit resultsChanged(false);
    }

    // Reverts your last filter or deletion step, restoring the previous list of found paths.
    void ResultsManager::undo() {
        if (!canUndo()) return;

        // Only push non-empty states to the redo history
        if (!m_currentResults.empty()) {
            m_redoStack.push(m_currentResults);
        }

        m_currentResults = m_undoStack.top();
        m_undoStack.pop();
        emit resultsChanged(false);
    }

    // Re-applies a filter or deletion step that you previously reverted.
    void ResultsManager::redo() {
        if (!canRedo()) return;

        // Only push non-empty states to the undo history
        if (!m_currentResults.empty()) {
            m_undoStack.push(m_currentResults);
        }

        m_currentResults = m_redoStack.top();
        m_redoStack.pop();
        emit resultsChanged(false);
    }

    void ResultsManager::clearHistory() {
        while (!m_undoStack.empty()) m_undoStack.pop();
        while (!m_redoStack.empty()) m_redoStack.pop();
    }

    void ResultsManager::clearSorting() {
        m_sortedColumnName = "";
        m_sortOrder = Qt::AscendingOrder;
        emit sortChanged("", Qt::AscendingOrder);
    }

    void ResultsManager::pushUndoState() {
        // Prevent empty results (the Dashboard state) from entering the Undo stack
        if (!m_currentResults.empty()) {
            m_undoStack.push(m_currentResults);
        }
        while (!m_redoStack.empty()) m_redoStack.pop();
    }

}
